### PR TITLE
I1-2901 Add support for selecting a map.

### PR DESCRIPTION
**Pull request by smpallen99**
Add support for proviiding a select that returns a map instaed to the schema structe for APIs that take select. 

I called the option `select_map`. 

Note, I also allowed it to be used with 1 field which will behave like `select` in that case, to avoid error handlineg.

### CODE REVIEW

File: lib/one_model.ex [hunk 1]
```diff
@@ -161,6 +161,9 @@ defmodule OneModel do
       * `select: atom | list`
         * Passing an atom will return the value of the given atom key
         * Passing a list of keys will return the default schema with the values for the specified keys.
+      * `select_map: atom | list`
+        * Passing an atom will return the value of the given atom key
```
Comment (Sinc63) [inline lib/one_model.ex]: I don't think this makes sense. If I want to get back the single value I can just pass `select: key`, so why not let this do some thing different from that, which is to be consistent with the rest of the select_map function and have it return `%{key: value}`.  
Either that or don't accept an atom and, but let me have a single entry list, which gets the same `{%key: value}` response I said above.

File: lib/one_model.ex [hunk 2]
```diff
@@ -243,6 +249,11 @@ defmodule OneModel do
       defp do_select(query, key) when is_atom(key), do: select(query, [b], field(b, ^key))
       defp do_select(query, keys), do: select(query, ^keys)
```
Comment (Sinc63) [inline lib/one_model.ex:249]: Your other option to provide this functionality would be to drop this line and put the `do` of line 255 here. I don't believe that anything is currently using `select: list`, partly because the return value is pretty ugly, and you could change that operation to return a map instead. Saves us an option, and eliminates an operation that is unlikely to ever be used, especially if this one is available.

In other words, don't implement a new option, just put the sensible part of this new operation in place of the rather ugly solution that is almost certainly not use for select: list.

File: lib/one_model.ex [hunk 3]
```diff
@@ -161,6 +161,9 @@ defmodule OneModel do
       * `select: atom | list`
         * Passing an atom will return the value of the given atom key
         * Passing a list of keys will return the default schema with the values for the specified keys.
+      * `select_map: atom | list`
```
Comment (palexanderm) [inline lib/one_model.ex]: I've got a different proposal. Rather than creating a `:select_map` option, let's expand upon the current `:select` option.
1.  Have it accept a map of fields of the form `%{field: 1, another_field: 1, some_other_field: 1}`. 
    - The map of fields to select would be consistent in format to the way query_sort_and_paginate fields are specified.
    - It could be called as such `list(select: :maps.from_keys([:a, :b], 1))`. 
    -  Could be extended to support fields to not select.
2. Similar to the above, but instead of map, accept a tuple of the type `{[atom], non_neg_integer}` where the integer is only either 0 or 1.
    - I only include this in case the first suggestion is too difficult to easily determine if the operation should be a select or reject. Although, I don't think it should be since we could more than likely reuse the logic from the query_sort_and_paginate selection.

File: lib/one_model.ex (hunk 1 above)
Comment (palexanderm) [reply lib/one_model.ex]: I don't agree that this option (if kept and my proposal is rejected) should allow for atoms to be passed. It contradicts the purpose and name of the option to "select a map".

File: lib/one_model.ex (hunk 2 above)
Comment (palexanderm) [reply lib/one_model.ex:249]: I did that in my original implementation of support the select option. Steve rejected it because it's inconsistent with Ecto's API. Passing the select option  a list of keys in Ecto  returns a schema. But I think my suggestion should be a good way around that.

**Review by palexanderm — CHANGES_REQUESTED** [review]

File: lib/one_model.ex (hunk 1 above)
Comment (smpallen99) [reply lib/one_model.ex]: Ok. I'll raise if only an atom is given.  Returing a map with 1 key seems silly, just pass [:key] for that.

Unless anyone has a better suggestion.

File: lib/one_model.ex (hunk 3 above)
Comment (smpallen99) [reply lib/one_model.ex]: @palexanderm As discussed: 

* The API query fields is an external API that makes sense. This PR is about an intenal API to be used by our code. They ary not the same thing.
* I can't think of a use case returning a map with all keys except those provede in the map `%{field1: 0, field2: 0}`. 
* Given that we will probably never use that option, `[:field1, :field2]` is more concise than `%{field1: 1, field2: 1}`. Furthermore, it aligns well with ecto, we just need to tell ecto if we want the struct returned (with default values) or a map. 
* I consided adding an option to the select key, but that just complicates select. A separate option is cleaner. 

So, I'll continue with the current design, but  remove the clause that acccepts an atom. So that will raise with a function clause error.

File: lib/one_model.ex (hunk 2 above)
Comment (smpallen99) [reply lib/one_model.ex:249]: Our goal  is, and should be, to be consistent with Ecto for internal use. 

Ecto has 2 options for select.
* select: [:one, :two] - returns the struct with only those 2 keys fetched. all other fields are the schema's defaults
* select: map(binding, [:one, :two]) - returns a map with just those 2 fields.

This is the behaviour I want in our app. So it's just a matter of the api that translates into those 2 options. 

The simplest and most concise approach is to have OneModel allow a new key for selecting a map. Which is what I've implemented.

See my reply on comment below for results of my conversation with Paul on this topic.
